In [1]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [2]:
def simplified_jaccard_similarity(doc1, doc2):
    tokens1 = doc1.lower().split()
    tokens2 = doc2.lower().split()

    lemmas1 = {token.rstrip('s') for token in tokens1}
    lemmas2 = {token.rstrip('s') for token in tokens2}

    intersection = lemmas1.intersection(lemmas2)
    union = lemmas1.union(lemmas2)
    similarity = len(intersection) / len(union) if union else 0
    return similarity

def calculate_similarity_for_pairs(pairs):
    for index, pair in enumerate(pairs, start=1):
        similarity = simplified_jaccard_similarity(pair[0], pair[1])
        print(f"Jaccard Similarity for pair {index}: {similarity:.2f}")

pairs = [
    ["What you do defines you", "Your deeds define you"],
    ["Once upon a time there lived a king.", "Who is your queen?"],
    ["He is desperate", "Is he not desperate?"]
]

calculate_similarity_for_pairs(pairs)


Jaccard Similarity for pair 1: 0.33
Jaccard Similarity for pair 2: 0.00
Jaccard Similarity for pair 3: 0.40


What is the Jaccard similarity index between pair1[0] and pair1[1]?


0.33